File: src/main/java/co/uk/isxander/lunarspoof/websocket/LunarAuthWebSocket.java
package co.uk.isxander.lunarspoof.websocket;

import co.uk.isxander.lunarspoof.LunarSpoof;
import co.uk.isxander.lunarspoof.websocket.packet.LunarPacket;
import co.uk.isxander.lunarspoof.websocket.packet.impl.CPacketEncryptionResponse;
import co.uk.isxander.lunarspoof.websocket.packet.impl.SPacketAuthenticatedRequest;
import co.uk.isxander.lunarspoof.websocket.packet.impl.SPacketEncryptionRequest;
import co.uk.isxander.xanderlib.utils.Constants;
import co.uk.isxander.xanderlib.utils.json.BetterJsonObject;
import com.mojang.authlib.exceptions.AuthenticationException;
import com.mojang.authlib.minecraft.MinecraftSessionService;
import com.mojang.authlib.yggdrasil.YggdrasilAuthenticationService;
import net.minecraft.util.CryptManager;
import org.java_websocket.client.WebSocketClient;
import org.java_websocket.drafts.Draft_6455;
import org.java_websocket.handshake.ServerHandshake;

import javax.crypto.SecretKey;
import java.math.BigInteger;
import java.net.Proxy;
import java.net.URI;
import java.net.URISyntaxException;
import java.nio.ByteBuffer;
import java.security.PublicKey;
import java.util.Map;
import java.util.UUID;
import java.util.function.Consumer;

public class LunarAuthWebSocket extends WebSocketClient implements Constants {

    private Consumer<String> consumer;
    private boolean whatIsThis = false;

    public LunarAuthWebSocket(Map<String, String> httpHeaders, Consumer<String> consumer) throws URISyntaxException {
        super(new URI("wss://authenticator.lunarclientprod.com"), new Draft_6455(), httpHeaders, 30000);
        this.consumer = consumer;
    }

    @Override
    public void onOpen(ServerHandshake handshakedata) {
        LunarSpoof.LOGGER.info("Connected to Lunar web socket.");
    }

    @Override
    public void onMessage(String message) {}

    @Override
    public void onMessage(ByteBuffer bytes) {
        processMessage(new BetterJsonObject(new String(bytes.array())));
    }

    public void processMessage(BetterJsonObject json) {
        LunarSpoof.LOGGER.info("Processing Message:\n" + json.toPrettyString());

        LunarPacket packet;
        String packetType = json.optString("packetType");
        switch (packetType) {
            case "SPacketEncryptionRequest":
                LunarSpoof.LOGGER.info("Packet: Encryption Request");

                packet = new SPacketEncryptionRequest();
                break;
            case "SPacketAuthenticatedRequest":
                LunarSpoof.LOGGER.info("Packet: Authentication Request");

                packet = new SPacketAuthenticatedRequest();
                break;
            default:
                return;
        }

        packet.processJson(json);
        packet.process(this);

    }

    public void acceptEncryption(SPacketEncryptionRequest packet) {
        SecretKey secretKey = CryptManager.createNewSharedKey();
        PublicKey publicKey = packet.getKey();
        byte[] keyHash = CryptManager.getServerIdHash("", publicKey, secretKey);
        if (keyHash == null) return;

        String str = (new BigInteger(keyHash)).toString(16);
        try {
            MinecraftSessionService session = new YggdrasilAuthenticationService(Proxy.NO_PROXY, UUID.randomUUID().toString()).createMinecraftSessionService();
            session.joinServer(mc.getSession().getProfile(), mc.getSession().getToken(), str);
        } catch (AuthenticationException ex) {
            ex.printStackTrace();
        } catch (NullPointerException npe) {
            close();
        }
        sendPacket(new CPacketEncryptionResponse(secretKey, publicKey, packet.getBytes()));
    }

    public void acceptAuthentication(SPacketAuthenticatedRequest packet) {
        this.whatIsThis = true;
        close();
        this.consumer.accept(packet.getJwtKey());
    }

    public void sendPacket(LunarPacket packet) {
        if (!isOpen()) return;

        BetterJsonObject json = new BetterJsonObject();
        json.addProperty("packetType", packet.getName());

        packet.processJson(json);
        send(json.toPrettyString());
    }

    @Override
    public void onClose(int code, String reason, boolean remote) {
        LunarSpoof.LOGGER.info(String.format("Connection Closed (%d, \"%s\")", code, reason));
        if (this.whatIsThis) return;
        this.consumer.accept(null);
    }

    @Override
    public void onError(Exception e) {
        e.printStackTrace();
    }

}

File: src/main/kotlin/dev/isxander/lunarspoof/websocket/auth/LunarAuthWebSocket.kt
package dev.isxander.lunarspoof.websocket.auth

import com.mojang.authlib.exceptions.AuthenticationException
import com.mojang.authlib.yggdrasil.YggdrasilAuthenticationService
import dev.isxander.lunarspoof.mc
import dev.isxander.lunarspoof.websocket.auth.packet.LunarPacket
import dev.isxander.lunarspoof.websocket.auth.packet.impl.CPacketEncryptionResponse
import dev.isxander.lunarspoof.websocket.auth.packet.impl.SPacketAuthenticatedRequest
import dev.isxander.lunarspoof.websocket.auth.packet.impl.SPacketEncryptionRequest
import gg.essential.api.utils.JsonHolder
import net.minecraft.util.CryptManager
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger
import org.java_websocket.client.WebSocketClient
import org.java_websocket.drafts.Draft_6455
import org.java_websocket.handshake.ServerHandshake
import java.math.BigInteger
import java.net.Proxy
import java.net.URI
import java.nio.ByteBuffer
import java.util.*


class LunarAuthWebSocket(
    httpHeaders: Map<String, String>,
    private val consumer: (String?) -> Unit
) : WebSocketClient(
    URI("wss://authenticator.lunarclientprod.com"),
    Draft_6455(),
    httpHeaders,
    30000
) {
    private var authenticated = false
    override fun onOpen(handshakedata: ServerHandshake) {
        LOGGER.info("Connected.")
    }

    override fun onMessage(message: String) {}
    override fun onMessage(bytes: ByteBuffer) {
        processMessage(JsonHolder(String(bytes.array())))
    }

    fun processMessage(json: JsonHolder) {
        val packet: LunarPacket
        val packetType = json.optString("packetType")
        packet = when (packetType) {
            "SPacketEncryptionRequest" -> SPacketEncryptionRequest()
            "SPacketAuthenticatedRequest" -> SPacketAuthenticatedRequest()
            else -> return
        }
        packet.processJson(json)
        packet.process(this)
        LOGGER.info("Processed Packet: " + packet.name)
    }

    fun acceptEncryption(packet: SPacketEncryptionRequest) {
        LOGGER.info("Accepting Encryption")
        val secretKey = CryptManager.createNewSharedKey()
        val publicKey = packet.key
        val keyHash = CryptManager.getServerIdHash("", publicKey, secretKey) ?: return
        val str = BigInteger(keyHash).toString(16)
        try {
            val session = YggdrasilAuthenticationService(
                Proxy.NO_PROXY,
                UUID.randomUUID().toString()
            ).createMinecraftSessionService()
            session.joinServer(mc.session.profile, mc.session.token, str)
        } catch (ex: AuthenticationException) {
            ex.printStackTrace()
        } catch (npe: NullPointerException) {
            close()
        }
        sendPacket(CPacketEncryptionResponse(secretKey, publicKey, packet.bytes))
    }

    fun acceptAuthentication(packet: SPacketAuthenticatedRequest) {
        authenticated = true
        close()
        consumer(packet.jwtKey)
    }

    fun sendPacket(packet: LunarPacket) {
        if (!isOpen) return
        LOGGER.info("Sending Packet: " + packet.name)
        val json = JsonHolder()
        json.put("packetType", packet.name)
        packet.processJson(json)
        send(json.toString())
    }

    override fun onClose(code: Int, reason: String, remote: Boolean) {
        if (code == 1000) {
            LOGGER.info("Authentication Completed.")
        }
        LOGGER.info(String.format("Connection Closed (%d, \"%s\")", code, reason))
        if (authenticated) return
        consumer(null)
    }

    override fun onError(e: Exception) {
        e.printStackTrace()
    }

    companion object {
        val LOGGER: Logger = LogManager.getLogger("LS Auth")
    }
}